Auth redirect (#130)
infinum/frontend-handbook: React/Recipes/Session Handling.md

## Motivation

In almost every application we create, there is some session. _What is saved in this session is not essential now._ The session needs to be accessible in our components and based on that data, we can restrict/allow users access to a page.

## `useSession` hook

To make a session accessible in the React component, we'll create a `useSession` hook that will expose a session.

### SWR example

```tsx
import { useCallback } from 'react';
import useSWR, { SWRConfiguration } from 'swr';

interface IUseAuthOptions extends SWRConfiguration {
  /**
   * on logout success callback
   */
  onLogout?(): void;
  /**
   * on login success callback
   */
  onLogin?(session: SessionModel): void;
}

export const useAuth = ({ onLogin, onLogout, ...config }: IUseAuthOptions = {}) => {

>   const datx = useDatx();

Reviewer: You are missing `useDatx` import

  const state = useSWR('current_session', () => readSession(), {
    shouldRetryOnError: false,
    errorRetryCount: 0,
    ...config,
  });

  const login = useCallback(
    async (attributes: ICreateSessionAttributes) => {
      const session = await state.mutate(createSession(attributes), false);

      if (onLogin) {
        onLogin(session);
      }

      return session;
    },
    [state, datx, onLogin]
  );

  const logout = useCallback(async () => {
    const session = await state.mutate(deleteSession(), false);

    if (onLogout) {
      onLogout();
    }

    return session;
  }, [state, datx, onLogout]);

  return { state, login, logout };
};
```

> This hook will expose session data and basic session handlers: `login` and `logout. Now, this hook can use this in any component to get session data.

Reviewer: ```suggestion
This hook will expose session data and basic session handlers: `login` and `logout`. Now, this hook can be used in any component to get session data.
```

}
```

Although this might look like a simple page redirect, render will occur and might hurt app performance. Once the page is loaded and the component is mounted, `state.data` will be `undefined`. Once request to read a session, `state.data` or `state.error` will be defined, and re-render will occur. This means that all child components will be re-rendered, too, which can potentially be an expensive operation. To optimize this, we'll create a component that will handle the redirect.

```tsx
interface IAuthRedirectProps {
  /**
   * URL used to redirect a user to
   * By default, if only this prop is set, the component will redirect if no session is found
   */
  to: Url | string;
  /**
   * If this property is set to `true`, user will be redirected if he is logged in.
   * Useful when you don't want to show login page to already logged in users.
   */
  ifFound?: boolean;
  /**
   * Callback that will trigger a redirect if true is returned.
   * Useful when you need to redirect base on some attribute, e.g. if user is not admin
   */
  condition?(session: SessionModel): boolean;
}

const AuthRedirect: FC<IAuthRedirectProps> = ({ to, ifFound, condition }) => {

>   const { state } = useAuth();

Reviewer: ```suggestion
  const { state: { data, isValidating, error } } = useAuth();
```

  useEffect(() => {
    // if state is validating, wait until request is done
    if (state.isValidating) {
      return;
    }

    // https://swr.vercel.app/advanced/performance#dependency-collection
    const hydration =
      state.data === undefined && state.error === undefined && state.isValidating === false;

    if (hydration) {
      return;
    }

    // condition has a priority over a ifFound property
    if (condition) {
      if (condition(state.data)) {
        router.push(to);
      }

      return;
    }

    const shouldRedirect = (ifFound && state.data) || (!ifFound && !state.data);

    if (shouldRedirect) {

>       router.push(to);

Reviewer: You are missing `useRouter` at the beginning of component

    }

>   }, [state.data, state.isValidating, state.error, to, condition, ifFound, router]);

Reviewer: ```suggestion
  }, [data, isValidating, error, to, condition, ifFound, router]);
```

  // this component renders nothing since it is only used to redirect if needed.
  return null;
};
```

If we now implement this in our example, this now looks like this.

```tsx
const SomePrivatePage = () => {
  return (

>     <>

Reviewer: You don't need Fragment here, and we agreed on `<Fragment>` syntax
